man/bcache.8.pod
=head1 NAME

bcache - Bolo Caching Daemon

=head1 SYNOPSIS

B<bcache> [options] -e tcp://bolo:2997 -B tcp://127.0.0.1:2997

=head1 DESCRIPTION

#INTRO

B<bcache> is a special bit of bolo infrastructure that subscribes to an upstream
B<bolo>(8) instance (the I<endpoint>), and caches all seen metric and state PDUs
while relaying them through to its own subscribers.  Additionally, it binds a control
port that allows clients to request a retransmission of all cached data, as if it were
being freshly broadcast, preserving timestamps.

This can be useful for stateless user tools that wish to have an up-to-date copy of
all known states and metrics, without having to store that state in a data store.

=head1 OPTIONS

=over

=item B<-V>, B<--version>

Print version and copyright information.

=item B<-e>, B<--endpoint> I<tcp://host:port>

The bolo broadcast endpoint.  Defaults to I<tcp://127.0.0.1:2997>.
Supports DNS resolution for both IPv4 and IPv6 endpoints.

=item B<-B>, B<--broadcast> I<tcp://*:port>

Where to re-broadcast from.  Clients connecting to this port can expect the
exact same protocol as the B<bolo>(8) broadcast endpoint, with the potential for
repeats and re-transmissions of older PDUs.

Defaults to I<tcp://127.0.0.1:2897>.

=item B<-l>, B<--listen> I<tcp://*:port>

What address and port to bind on for the controller channel.  Clients can connect
to this (via a PUSH socket) and submit a B<REPEAT> PDU to cause a re-broadcast of all
known PDUs to all connected clients.

Defaults to I<tcp://127.0.0.1:2898>.

=item B<-F>, B<--foreground>

By default, B<bcache> will fork into the background, detach its terminal
and daemonize itself.  This option inhibits that behavior, and also stops
its from setting the effective UID / GID (see B<-u> and B<-g>).

=item B<-p>, B<--pidfile> I</path/to/pidfile>

Specify where B<bcache> should write its PID to, for control by init
scripts.  Defaults to B</var/run/bcache.pid>.  Has no effect if B<-F> is
given.

=item B<-u>, B<--user> I<USERNAME>

=item B<-g>, B<--group> I<GROUP NAME>

User and group to drop privileges to.  By default, B<bcache> will run as
root:root, which is probably not what you want.

=item B<-v>, B<--verbose>

Enable verbose mode, printing debugging information to standard error.

=back

=head1 TRANSPARENT PROXY MODE

If you run B<bolo>(8) and bind its broadcast endpoint to something other
than I<tcp://*:2997> (say, for instance, I<tcp://127.0.0.1:1234>),
you can transparently front it with B<bcache> broadcasting on I<tcp://*:2997>
and clients will never know the difference.

=head1 SEE ALSO

#SEEALSO

=head1 AUTHOR

#AUTHOR


debian/changelog
bolo (0.2.3-1) precise; urgency=low

  * Stub changelog, built by changelog-generator

 -- Package Builder <packages@niftylogic.com>  Tue, 19 May 2015 22:09:20 -0400


src/bcache.c
/*
  Copyright 2015 James Hunt <james@jameshunt.us>

  This file is part of Bolo.

  Bolo is free software: you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.

  Bolo is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along
  with Bolo.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "bolo.h"
#include "subs.h"
#include <pthread.h>
#include <signal.h>
#include <getopt.h>
#include <assert.h>

/**************************/

typedef struct {
	hash_t pdus;      /* map of PDUs, indexed by name, that we've seen */

	void *control;    /* SUB:  hooked up to supervisor.command; receives control messages */

	void *subscriber; /* SUB:  subscriber to bolo broadcast port (external) */
	void *broadcast;  /* PUB:  bolo re-broadcast port (external) */
	void *listen;     /* PULL: local listener for prompted re-broadcast requests (external) */

	reactor_t *reactor;
} dispatcher_t;

int dispatcher_thread(void *zmq, const char *endpoint, const char *listen, const char *broadcast);

static int _dispatcher_reactor(void *socket, pdu_t *pdu, void *_) /* {{{ */
{
	assert(socket != NULL);
	assert(pdu != NULL);
	assert(_ != NULL);

	dispatcher_t *dispatcher = (dispatcher_t*)_;

	if (socket == dispatcher->control)
		return VIGOR_REACTOR_HALT;

	if (socket == dispatcher->subscriber) {
		char *key = NULL;

		/* re-broadcast immediately! */
		logger(LOG_DEBUG, "relaying [%s] PDU", pdu_type(pdu));
		pdu_send_and_free(pdu_dup(pdu, NULL), dispatcher->broadcast);

		if (strcmp(pdu_type(pdu), "COUNTER") == 0
		 || strcmp(pdu_type(pdu), "SAMPLE") == 0
		 || strcmp(pdu_type(pdu), "RATE") == 0) {
			key = pdu_string(pdu, 2);

		} else if (strcmp(pdu_type(pdu), "STATE") == 0) {
			key = pdu_string(pdu, 1);
		}

		if (key) {
			logger(LOG_DEBUG, "storing %s PDU for %s for re-broadcast later", pdu_type(pdu), key);

			pdu_t *dupe = pdu_dup(pdu, NULL);
			pdu_t *prev = hash_set(&dispatcher->pdus, key, dupe);
			if (prev != dupe)
				pdu_free(prev);

			free(key);
		}

		return VIGOR_REACTOR_CONTINUE;
	}

	if (socket == dispatcher->listen) {
		if (strcmp(pdu_type(pdu), "REPEAT") == 0) {
			logger(LOG_INFO, "received a [REPEAT] request");
			char *k;
			pdu_t *dupe;

			int n = 0;
			for_each_key_value(&dispatcher->pdus, k, dupe) {
				n++;
				logger(LOG_DEBUG, "re-broadcasting %s for %s", pdu_type(dupe), k);
				pdu_send_and_free(pdu_dup(dupe, NULL), dispatcher->broadcast);
			}
			logger(LOG_INFO, "re-broadcast %i PDUs to all connected clients", n);

		} else {
			logger(LOG_ERR, "received unrecognized [%s] PDU on listener", pdu_type(pdu));
		}

		return VIGOR_REACTOR_CONTINUE;
	}

	logger(LOG_ERR, "unhandled socket!");
	return VIGOR_REACTOR_HALT;
}
/* }}} */
static void * _dispatcher_thread(void *_) /* {{{ */
{
	assert(_ != NULL);

	dispatcher_t *dispatcher = (dispatcher_t*)_;

	reactor_go(dispatcher->reactor);

	logger(LOG_DEBUG, "dispatcher: shutting down");

	zmq_close(dispatcher->broadcast);
	zmq_close(dispatcher->listen);
	zmq_close(dispatcher->subscriber);
	zmq_close(dispatcher->control);

	reactor_free(dispatcher->reactor);
	free(dispatcher);

	logger(LOG_DEBUG, "dispatcher: terminated");
	return NULL;
}
/* }}} */
int dispatcher_thread(void *zmq, const char *endpoint, const char *listen, const char *broadcast) /* {{{ */
{
	assert(zmq != NULL);
	assert(endpoint != NULL);
	assert(listen != NULL);
	assert(broadcast != NULL);

	int rc;

	logger(LOG_INFO, "initializing dispatcher thread");

	dispatcher_t *dispatcher = vmalloc(sizeof(dispatcher_t));

	logger(LOG_DEBUG, "connecting dispatcher.control -> supervisor");
	rc = subscriber_connect_supervisor(zmq, &dispatcher->control);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "connecting dispatcher.subscriber <- bolo at %s", endpoint);
	dispatcher->subscriber = zmq_socket(zmq, ZMQ_SUB);
	if (!dispatcher->subscriber)
		return -1;
	rc = zmq_setsockopt(dispatcher->subscriber, ZMQ_SUBSCRIBE, "", 0);
	if (rc != 0)
		return rc;
	rc = vzmq_connect_af(dispatcher->subscriber, endpoint, AF_UNSPEC);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "binding PULL socket dispatcher.listen to %s", listen);
	dispatcher->listen = zmq_socket(zmq, ZMQ_PULL);
	if (!dispatcher->listen)
		return -1;
	rc = zmq_bind(dispatcher->listen, listen);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "binding PUB socket dispatcher.broadcast to %s", broadcast);
	dispatcher->broadcast = zmq_socket(zmq, ZMQ_PUB);
	if (!dispatcher->broadcast)
		return -1;
	rc = zmq_bind(dispatcher->broadcast, broadcast);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "setting up dispatcher event reactor");
	dispatcher->reactor = reactor_new();
	if (!dispatcher->reactor)
		return -1;

	logger(LOG_DEBUG, "dispatcher: registering dispatcher.control with event reactor");
	rc = reactor_set(dispatcher->reactor, dispatcher->control, _dispatcher_reactor, dispatcher);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "dispatcher: registering dispatcher.subscriber with event reactor");
	rc = reactor_set(dispatcher->reactor, dispatcher->subscriber, _dispatcher_reactor, dispatcher);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "dispatcher: registering dispatcher.listen with event reactor");
	rc = reactor_set(dispatcher->reactor, dispatcher->listen, _dispatcher_reactor, dispatcher);
	if (rc != 0)
		return rc;

	logger(LOG_DEBUG, "dispatcher: spinning up thread");
	pthread_t tid;
	rc = pthread_create(&tid, NULL, _dispatcher_thread, dispatcher);
	if (rc != 0)
		return rc;

	return 0;
}
/* }}} */

/**************************/

int main(int argc, char **argv) /* {{{ */
{
	struct {
		char *endpoint;
		char *listen;
		char *broadcast;

		int   verbose;
		int   daemonize;

		char *pidfile;
		char *user;
		char *group;

	} OPTIONS = {
		.verbose   = 0,
		.endpoint  = strdup("tcp://127.0.0.1:2997"),
		.listen    = strdup("tcp://127.0.0.1:2898"),
		.broadcast = strdup("tcp://127.0.0.1:2897"),
		.daemonize = 1,
		.pidfile   = strdup("/var/run/bcache.pid"),
		.user      = strdup("root"),
		.group     = strdup("root"),
	};

	struct option long_opts[] = {
		{ "help",             no_argument, NULL, 'h' },
		{ "version",          no_argument, NULL, 'V' },
		{ "verbose",          no_argument, NULL, 'v' },
		{ "endpoint",   required_argument, NULL, 'e' },
		{ "broadcast",  required_argument, NULL, 'B' },
		{ "listen",     required_argument, NULL, 'l' },
		{ "foreground",       no_argument, NULL, 'F' },
		{ "pidfile",    required_argument, NULL, 'p' },
		{ "user",       required_argument, NULL, 'u' },
		{ "group",      required_argument, NULL, 'g' },
		{ 0, 0, 0, 0 },
	};
	for (;;) {
		int idx = 1;
		int c = getopt_long(argc, argv, "h?Vv+e:Fp:u:g:S:P:l:B:", long_opts, &idx);
		if (c == -1) break;

		switch (c) {
		case 'h':
		case '?':
			printf("bcache v%s\n", BOLO_VERSION);
			printf("Usage: bcache [-h?FVv] [-e tcp://host:port] [-P PREFIX] [-S tcp://host:port]\n"
			       "                [-l tcp://host:port] [-B tcp://host:port]\n"
			       "                [-u user] [-g group] [-p /path/to/pidfile]\n\n");
			printf("Options:\n");
			printf("  -?, -h               show this help screen\n");
			printf("  -F, --foreground     don't daemonize, stay in the foreground\n");
			printf("  -V, --version        show version information and exit\n");
			printf("  -v, --verbose        turn on debugging, to standard error\n");
			printf("  -e, --endpoint       bolo broadcast endpoint to connect to\n");
			printf("  -B, --broadcast      address to bind and re-broadcast from\n");
			printf("  -l, --listen         address to listen on for re-broadcast requests\n");
			printf("  -u, --user           user to run as (if daemonized)\n");
			printf("  -g, --group          group to run as (if daemonized)\n");
			printf("  -p, --pidfile        where to store the pidfile (if daemonized)\n");
			exit(0);

		case 'V':
			printf("bcache v%s\n"
			       "Copyright (C) 2015 James Hunt\n",
			       BOLO_VERSION);
			exit(0);

		case 'v':
			OPTIONS.verbose++;
			break;

		case 'e':
			free(OPTIONS.endpoint);
			OPTIONS.endpoint = strdup(optarg);
			break;

		case 'F':
			OPTIONS.daemonize = 0;
			break;

		case 'p':
			free(OPTIONS.pidfile);
			OPTIONS.pidfile = strdup(optarg);
			break;

		case 'u':
			free(OPTIONS.user);
			OPTIONS.user = strdup(optarg);
			break;

		case 'g':
			free(OPTIONS.group);
			OPTIONS.group = strdup(optarg);
			break;

		case 'l':
			free(OPTIONS.listen);
			OPTIONS.listen = strdup(optarg);
			break;

		case 'B':
			free(OPTIONS.broadcast);
			OPTIONS.broadcast = strdup(optarg);
			break;

		default:
			fprintf(stderr, "unhandled option flag %#02x\n", c);
			exit(1);
		}
	}

	if (OPTIONS.daemonize) {
		log_open("bcache", "daemon");
		log_level(LOG_ERR + OPTIONS.verbose, NULL);

		mode_t um = umask(0);
		if (daemonize(OPTIONS.pidfile, OPTIONS.user, OPTIONS.group) != 0) {
			fprintf(stderr, "daemonization failed: (%i) %s\n", errno, strerror(errno));
			exit(3);
		}
		umask(um);
	} else {
		log_open("bcache", "console");
		log_level(LOG_INFO + OPTIONS.verbose, NULL);
	}
	logger(LOG_NOTICE, "starting up");


	void *zmq = zmq_ctx_new();
	if (!zmq) {
		logger(LOG_ERR, "failed to initialize 0MQ context");
		exit(1);
	}

	int rc;
	rc = subscriber_init();
	if (rc != 0) {
		logger(LOG_ERR, "failed to initialize subscriber architecture");
		exit(2);
	}

	rc = dispatcher_thread(zmq, OPTIONS.endpoint, OPTIONS.listen, OPTIONS.broadcast);
	if (rc != 0) {
		logger(LOG_ERR, "failed to initialize dispatcher: %s", strerror(errno));
		exit(2);
	}

	subscriber_supervisor(zmq);
	logger(LOG_INFO, "shutting down");

	free(OPTIONS.endpoint);

	free(OPTIONS.pidfile);
	free(OPTIONS.user);
	free(OPTIONS.group);

	return 0;
}
/* }}} */
